Guide format (#1)
mesutcelik/hazelcast-hpa: doc/modules/ROOT/pages/index.adoc

:hazelcast: Hazelcast IMDG
:framework: Kubernetes

= Kubernetes HPA for Hazelcast cluster

Cloud technologies give you on-demand options so that you can create compute, disk or network resources based on your requirements. When your demand changes, you update your infrastructure by releasing some resources or adding more. That is actually named Manual Scaling based on human intervention. Kubernetes is no different in this particular use case. If you create a service on top of Kubernetes, and you see unexpected popularity than you planned for, you need to scale out your number of Pods to match the traffic coming to your application. Kubernetes has an automated solution for this problem. https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/[Horizontal Pod Autoscaler (HPA)] automatically scales the number of pods in Kubernetes based on your metrics selection. You have two options to choose.

* https://kubernetes.io/docs/tasks/debug-application-cluster/resource-usage-monitoring/#resource-metrics-pipeline[Resource Metrics]
* https://kubernetes.io/docs/tasks/debug-application-cluster/resource-usage-monitoring/#full-metrics-pipeline[Custom Metrics]

> Let's dig into each of them with a supporting example using Hazelcast

Reviewer: I think we could shorten this paragraph to something like: `This guide will show you how to set up Hazelcast with [Kubernetes Horizontal Pod Autoscaler(HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/).`. I don't think this whole description is really needed.

include::{templates-url}/link-to-repo.adoc[]

== What You’ll Learn

> In this guide you will setup deploy Hazelcast to Kubernetes cluster and set up HPA using Resource Metrics and Custom Metrics.

Reviewer 2: > you will setup deploy Hazelcast to Kubernetes

could be `setup Hazelcast deployment to`? the sentence sounds broken to me.

== Prerequisites

* ~15 minutes
* Kubernetes cluster with kubectl configured
* https://helm.sh/[Helm 3] tool installed

== Installing Hazelcast Helm Chart

Helm is the package manager for Kubernetes, and we will use it throughout the document to install various software. This is the https://github.com/helm/helm#install[link] to install `helm` into your computer. Once you have a working `Helm 3` we can install https://github.com/hazelcast/charts/tree/master/stable/hazelcast[Hazelcast Helm Chart] by running the following commands:

[source, shell]
----
$ helm repo add hazelcast https://hazelcast-charts.s3.amazonaws.com/
$ helm repo update
$ helm install hazelcast hazelcast/hazelcast
----

> Not you can verify that Hazelcast cluster is successfully deployed:

Reviewer 2: ```suggestion
Now you can verify that Hazelcast cluster is successfully deployed:
```

[source, shell]
----
$ kubectl get all
NAME                        READY   STATUS    RESTARTS   AGE
pod/hazelcast-0             1/1     Running   0          7m35s
pod/hazelcast-1             1/1     Running   0          7m3s
pod/hazelcast-2             1/1     Running   0          6m23s
pod/hazelcast-mancenter-0   1/1     Running   0          7m35s

NAME                          TYPE           CLUSTER-IP      EXTERNAL-IP    PORT(S)                        AGE
service/hazelcast             ClusterIP      None            <none>         5701/TCP                       7m35s
service/hazelcast-mancenter   LoadBalancer   10.175.245.87   34.77.64.138   8080:30957/TCP,443:30042/TCP   7m35s
service/kubernetes            ClusterIP      10.175.240.1    <none>         443/TCP                        25h

NAME                                   READY   AGE
statefulset.apps/hazelcast             3/3     7m35s
statefulset.apps/hazelcast-mancenter   1/1     7m35s
----

> For more details on installing using Hazelcast Helm Chart follow the https://hazelcast.com/blog/hazelcast-helm-chart/[link].

Reviewer 2: nit
```suggestion
For more details on installing and using Hazelcast Helm Chart you can see this https://hazelcast.com/blog/hazelcast-helm-chart/[blogpost].
```

== Horizontal Pod AutoScaler (Resource Metrics)

As you can read the details from official Kubernetes documentation, https://kubernetes.io/docs/tasks/debug-application-cluster/resource-usage-monitoring/#resource-metrics-pipeline[Resource Metrics] provide CPU and Memory based metrics for Pods and Nodes in your Kubernetes Cluster. Those metrics are exposed via the metrics.k8s.io API and one implementation of that API is Metrics Server.

> Before moving forward, verify that Metrics Server is properly installed and in the list of API Registration.

Reviewer: ```suggestion
Before moving forward, verify that Metrics Server is properly installed and visible in the list of API Registration.
```

[source, shell]
----
$ kubectl get apiservices.apiregistration.k8s.io | grep metrics-server
v1beta1.metrics.k8s.io                 kube-system/metrics-server   True        26h
----

Let's create an HPA based on CPU usage.

[source, shell]
----
$ kubectl autoscale statefulset hazelcast --cpu-percent=50 --min=3 --max=10
----

> This HPA will periodically check Hazelcast StatefulSet CPU usage and will decide number of running pods between 3 and 10, based on some https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details[calculations]. The simplest way to put some cpu load on Hazelcast Pod is executing https://en.wikipedia.org/wiki/Yes_(Unix)[yes tool]. This is just to show how HPA is triggered to scale up Hazelcast Cluster by printing yes in one of Hazelcast Pods. You should use a proper load testing tool to test HPA in your Hazelcast Cluster. Before generating CPU load, you can open 2 new terminals to watch HPA target values and number of hazelcast pods via watch `kubectl get pods` and watch `kubectl get hpa` commands. Let's move on and execute following command for 5-10 seconds and terminate via *Ctrl + C*.

Reviewer: ```suggestion
This HPA will periodically check Hazelcast StatefulSet CPU usage and will decide the number of running pods between 3 and 10, based on some https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details[calculations]. The simplest way to put some cpu load on Hazelcast Pod is to execute https://en.wikipedia.org/wiki/Yes_(Unix)[yes tool]. This is just to show how HPA is triggered to scale up Hazelcast Cluster by printing yes in one of Hazelcast Pods. You should use a proper load testing tool to test HPA in your Hazelcast Cluster. Before generating CPU load, you can open 2 new terminals to watch HPA target values and number of hazelcast pods via watch `kubectl get pods` and watch `kubectl get hpa` commands. Let's move on and execute the following command for 5-10 seconds and terminate via *Ctrl + C*.
```

> You should see now that your HPA target is above 50%, and some new pods are started. As initial hazelcast cluster was 3 members cluster, `hazelcast-3` and above are new pods created by HPA.

Reviewer 2: nit
```suggestion
Now that your HPA target is above 50%, you should see some new pods have been created. As the initial Hazelcast cluster was 3 members cluster, `hazelcast-3` and above are the new pods created by HPA.
```

[source, shell]
----
$ kubectl get hpa
NAME             REFERENCE                  TARGETS     MINPODS   MAXPODS   REPLICAS   AGE
hazelcast        StatefulSet/hazelcast      282m/200m   3         5         5          18m

$ kubectl get pods
NAME                                    READY   STATUS        RESTARTS   AGE
hazelcast-0                             1/1     Running       0          30m
hazelcast-1                             1/1     Running       0          30m
hazelcast-2                             1/1     Running       0          29m
hazelcast-3                             1/1     Running       0          14m
hazelcast-4                             1/1     Running       0          13m
hazelcast-mancenter-0                   1/1     Running       1          30m
----

=== Cleanup

As you successfully managed to use Resource Metrics with Hazelcast, we can clean up resources used up to that point.

[source, shell]
----
$ helm uninstall hazelcast
release "hazelcast" uninstalled

$ kubectl delete hpa hazelcast
horizontalpodautoscaler.autoscaling "hazelcast" deleted
----

== Custom Metrics

> In the previous section, we have explained how to use Resource Metrics to autoscale your deployments based on CPU or Memory Metrics. Although that is fine for some architectures, those metrics are Kubernetes Pod or Node level so application level autoscaling is not possible with Resource Metrics. Kubernetes introduced Custom Metrics API in order to fill in this gap. When using Custom Metrics API, each container exposes their own metrics and HPA uses those metrics to make autoscaling decisions. In this example, we will use https://prometheus.io/[Prometheus] as Metrics Storage and https://github.com/DirectXMan12/k8s-prometheus-adapter[Prometheus Adapter] as the Custom Metrics API Provider.

Reviewer: ```suggestion
In the previous section, we have explained how to use Resource Metrics to autoscale your deployments based on CPU or Memory Metrics. Although that is fine for some architectures, those metrics are Kubernetes Pod or Node level so application-level autoscaling is not possible with Resource Metrics. Kubernetes introduced Custom Metrics API in order to fill in this gap. When using Custom Metrics API, each container exposes its own metrics and HPA uses those metrics to make autoscaling decisions. In this example, we will use https://prometheus.io/[Prometheus] as Metrics Storage and https://github.com/DirectXMan12/k8s-prometheus-adapter[Prometheus Adapter] as the Custom Metrics API Provider.
```

=== Install Prometheus and Prometheus Adapter

To install Prometheus, execute the following commands:

[source, shell]
----
$ helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
"prometheus-community" has been added to your repositories

$ kubectl create ns monitoring
namespace/monitoring created

$ helm install prometheus prometheus-community/prometheus --namespace=monitoring
----

> Before installing Prometheus Adapter let's take a look at the content of `hazelcast-custom-metrics-hpa.yaml`

Reviewer 2: isn't it `prometheus-adapter-values.yaml` content?

----

This configuration will be passed to the helm chart while deploying Prometheus Adapter but let's just go through each part before doing that. The config basically tells Prometheus Adapter:

* query only `jvm_memory_bytes_used` and `jvm_memory_bytes_max`
* assign `kubernetes_*` based labels to resources to be able to query via REST URLs like `/apis/custom.metrics.k8s.io/v1beta1/namespaces/default/services/*/on_heap_ratio`
* give a new, easier name (`on_heap_ratio`) to the metric that we expose via custom metrics adapter
* select max value out of all series provided by all PODs

This example uses `max` function while creating `metricsQuery`, but you can basically use some other https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators[aggregation operator] like `avg` in your own configuration. If you saved the file above, you can create a prometheus adapter based on that configuration.

To install Prometheus Adapter run the following command:

[source, shell]
----
$ helm install prometheus-adpter prometheus-community/prometheus-adapter --namespace=monitoring -f prometheus-adapter-values.yaml
----

=== Install Metrics Enabled Hazelcast Cluster

> Let's install a new 3 members hazelcast cluster with metrics enabled. Each hazelcast member container in this new deployment will expose their own metrics data under `/metrics` endpoint. This endpoint exposes metrics in Prometheus format because each hazelcast container is started with Prometheus https://github.com/prometheus/jmx_exporter[JMX Exporter]. This is a feature provided by https://github.com/hazelcast/hazelcast-docker[Hazelcast Docker] Image. We also set `resources.limits.memory=512Mi` which sets each hazelcast member JVM max heap size to `128Mi`. JVM by default grabs 25% of available memory as max heap size.

Reviewer 2: nit
```suggestion
Let's install a new 3 members Hazelcast cluster with metrics enabled. Each Hazelcast member container in this new deployment will expose their own metrics data under `/metrics` endpoint. This endpoint exposes metrics in Prometheus format because each Hazelcast container is started with Prometheus https://github.com/prometheus/jmx_exporter[JMX Exporter]. This is a feature provided by https://github.com/hazelcast/hazelcast-docker[Hazelcast Docker] Image. We also set `resources.limits.memory=512Mi` which sets each Hazelcast member JVM max heap size to `128Mi`. JVM by default grabs 25% of available memory as max heap size.
```

        metric:
          name: on_heap_ratio
        target:
          type: Value
          value: 200m
----

Apply HPA to your cluster with `kubectl`:

[source, shell]
----
$ kubectl apply -f hazelcast-custom-metrics-hpa.yaml
horizontalpodautoscaler.autoscaling/heap-based-hpa created
----

=== Generate some Memory Load for HPA

Let's have a look `TARGETS` part of HPA output

[source, shell]
----
$ kubectl get hpa heap-based-hpa
NAME             REFERENCE               TARGETS     MINPODS   MAXPODS   REPLICAS   AGE
heap-based-hpa   StatefulSet/hazelcast   180m/200m   3         10        10         11m
----

As we see, current HPA Target is 180m/200m so if we increase memory usage just 10%  by adding 10MB into the cluster, HPA should trigger a scale up event. I will use Hazelcast Java Client to put some data into cluster, but you can use your own language to implement the same functionality. You can see all Hazelcast supported programming languages https://hazelcast.org/imdg/clients-languages/[here].

Let's first port forward from our local machine to be able to connect remote k8s hazelcast member pod.

[source, shell]
----
$ kubectl port-forward hazelcast-0 5701
----

> Execute following code snippet to put data into Hazelcast Cluster.

Reviewer 2: the snippet is just a client instantiation. Let's either expand the snippet or indicate that this is the config needed to connect to the cluster and data should be inserted explicitly.

[source, java]
----

>  // start Hazelcast Client with smartRouting enabled

Reviewer 2: but we disable this 2 lines below.

 ClientConfig cfg = new ClientConfig();
 cfg.getNetworkConfig().setSmartRouting(false);
 HazelcastInstance client = HazelcastClient.newHazelcastClient(cfg);
----

[NOTE]
====
You can use the client application from xref:kubernetes-external-client:ROOT:index.adoc[External Smart Client] guide to connect to Hazelcast cluster.
====

> When you start putting data into your hazelcast cluster, you will see that new pods will be created and added to Hazelcast Cluster.

Reviewer 2: nit
```suggestion
When you start putting data into your Hazelcast cluster, you will see that new pods will be created and added to Hazelcast Cluster.
```

[source, shell]
----
$ kubectl get po
NAME                    READY   STATUS    RESTARTS   AGE
hazelcast-0             1/1     Running   0          73m
hazelcast-1             1/1     Running   0          72m
hazelcast-2             1/1     Running   0          72m
hazelcast-3             1/1     Running   0          16m
hazelcast-4             1/1     Running   0          3m
hazelcast-5             1/1     Running   0          3m
hazelcast-6             1/1     Running   0          47s
hazelcast-7             1/1     Running   0          41s
hazelcast-8             1/1     Running   0          14s
hazelcast-9             1/1     Running   0          13s
hazelcast-mancenter-0   1/1     Running   0          73m
----

== Conclusion

> Autoscaling is an important feature for enterprises to save money and to cope with unexpected traffic coming to your deployments. However, configuring autoscaling needs to be done carefully because you might end up unnecessary scale up/down operations which might cost some instability in your system. In this blog post, we explained how you can use HPA with your Hazelcast Cluster based on Resource Metrics and Custom Metrics. If Kubernetes Pod/Node Level Cpu/Memory usage is fine for you then use Resource Metrics. If you have more specific requirements, and you need to have Hazelcast specific autoscaling capabilities, Custom Metrics is the answer.

Reviewer 2: seems like this is written before guides platform comes in :)
```suggestion
Autoscaling is an important feature for enterprises to save money and to cope with unexpected traffic coming to your deployments. However, configuring autoscaling needs to be done carefully because you might end up unnecessary scale up/down operations which might cost some instability in your system. In this guide, we explained how you can use HPA with your Hazelcast Cluster based on Resource Metrics and Custom Metrics. If Kubernetes Pod/Node Level Cpu/Memory usage is fine for you then use Resource Metrics. If you have more specific requirements, and you need to have Hazelcast specific autoscaling capabilities, Custom Metrics is the answer.
```

== Software Versions

This is the list of software versions used in this guide.

[source, shell]
----

> elm ls --all-namespaces

Reviewer 2: ```suggestion
$ helm ls --all-namespaces
```
